packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+    }
+
+    private fun source(source: Source): Source {
+      return object : ForwardingSource(source) {
+        var totalBytesRead = 0L
+
+        @Throws(IOException::class)
+        override fun read(sink: Buffer, byteCount: Long): Long {
+          val bytesRead = super.read(sink, byteCount)
+          // read() returns the number of bytes read, or -1 if this source is exhausted.
+          totalBytesRead += if (bytesRead != -1L) bytesRead else 0
+          progressListener.update(
+            totalBytesRead,
+            responseBody?.contentLength()
+              ?: -1,
+            bytesRead == -1L
+          )
+          return bytesRead
+        }
+      }
+    }
+  }
+
+  internal interface ProgressListener {
+    fun update(bytesRead: Long, contentLength: Long, done: Boolean)
+  }
+
+  @get:Synchronized
+  private val okHttpClient: OkHttpClient?
+    get() {
+      if (client == null) {
+        val builder = OkHttpClient.Builder()
+          .connectTimeout(60, TimeUnit.SECONDS)
+          .readTimeout(60, TimeUnit.SECONDS)
+          .writeTimeout(60, TimeUnit.SECONDS)
+        val cookieHandler: CookieHandler by moduleRegistry()
+        builder.cookieJar(JavaNetCookieJar(cookieHandler))
+        client = builder.build()
+      }
+      return client
+    }
+
+  @Throws(IOException::class)
+  private fun md5(file: File): String {
+    val `is`: InputStream = FileInputStream(file)
+    return `is`.use {
+      val md5bytes = DigestUtils.md5(it)
+      String(Hex.encodeHex(md5bytes))
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun ensureDirExists(dir: File) {
+    if (!(dir.isDirectory || dir.mkdirs())) {
+      throw IOException("Couldn't create directory '$dir'")
+    }
+  }
+
+  /**
+   * Concatenated copy of org.apache.commons.io@1.4.0#FileUtils#forceDelete
+   * Newer version of commons-io uses File#toPath() under the hood that unsupported below Android SDK 26
+   * See docs for reference https://commons.apache.org/proper/commons-io/javadocs/api-1.4/index.html
+   */
+  @Throws(IOException::class)
+  private fun forceDelete(file: File) {
+    if (file.isDirectory) {
+      val files = file.listFiles() ?: throw IOException("Failed to list contents of $file")
+      var exception: IOException? = null
+      for (f in files) {
+        try {
+          forceDelete(f)
+        } catch (ioe: IOException) {
+          exception = ioe
+        }
+      }
+      if (null != exception) {
+        throw exception
+      }
+      if (!file.delete()) {
+        throw IOException("Unable to delete directory $file.")
+      }
+    } else if (!file.delete()) {
+      throw IOException("Unable to delete file: $file")
+    }
+  }
+
+  private fun getFileSize(file: File): Long {
+    if (!file.isDirectory) {
+      return file.length()
+    }
+    val content = file.listFiles() ?: return 0
+    var size: Long = 0
+    for (item in content) {
+      size += getFileSize(item)
+    }
```

Comment:
If you want a functional way (maybe you don't 😉 ):
```suggestion
    val size = content.map { getFileSize(it) }.reduceOrNull { total, itemSize -> total + itemSize } ?: 0
```
the `orNull` is because the collection can be empty, and `reduce` would throw in that case